File: app/src/main/java/com/friendlypos/distribucion/fragment/DistSelecClienteFragment.java
package com.friendlypos.distribucion.fragment;

import android.support.annotation.Nullable;
import android.os.Bundle;
import android.support.v4.view.MenuItemCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.friendlypos.R;
import com.friendlypos.distribucion.activity.DistribucionActivity;
import com.friendlypos.distribucion.adapters.DistrClientesAdapter;
import com.friendlypos.distribucion.adapters.DistrResumenAdapter;
import com.friendlypos.distribucion.modelo.Inventario;
import com.friendlypos.distribucion.modelo.Pivot;
import com.friendlypos.distribucion.modelo.sale;
import com.friendlypos.principal.modelo.Clientes;
import com.friendlypos.principal.modelo.Productos;

import java.util.ArrayList;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import io.realm.Realm;
import io.realm.RealmQuery;
import io.realm.RealmResults;
import io.realm.Sort;

import static io.realm.internal.SyncObjectServerFacade.getApplicationContext;


public class DistSelecClienteFragment extends BaseFragment  implements SearchView.OnQueryTextListener{
    private Realm realm;

    @Bind(R.id.recyclerViewDistrCliente)
    public RecyclerView recyclerView;

    private DistrClientesAdapter adapter;
    private DistrResumenAdapter adapter2;

    public static DistSelecProductoFragment getInstance() {
        return new DistSelecProductoFragment();
    }
    @Override
    public void onDestroy() {
        super.onDestroy();
    }
    @Override
    public void onResume() {
        super.onResume();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_distribucion_cliente, container,
                false);
        setHasOptionsMenu(true);
        ButterKnife.bind(this, rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if(adapter == null) {
            adapter = new DistrClientesAdapter(getContext(), ((DistribucionActivity) getActivity()), getListClientes());
            adapter2 = new DistrResumenAdapter();
        }
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.setAdapter(adapter);
        Log.d("listaProducto", getListClientes() + "");
    }

    private List<sale> getListClientes(){

        realm = Realm.getDefaultInstance();
        final RealmQuery<sale> query = realm.where(sale.class).equalTo("aplicada", 0).equalTo("devolucion", 0);
       // final RealmResults<sale> result1 = query.findAll();
        final RealmResults<sale> result1 = query.findAllSorted("created_at", Sort.DESCENDING);

        if (result1.isEmpty()) {

            Toast.makeText(getApplicationContext(),"Favor descargar datos primero",Toast.LENGTH_LONG).show();
        }

        return result1;

    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
       // realm.close();
    }

    @Override
    public void updateData() {
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.menu_main, menu);

        final MenuItem item = menu.findItem(R.id.action_search);
        final SearchView searchView = (SearchView) MenuItemCompat.getActionView(item);
        searchView.setOnQueryTextListener(this);

        MenuItemCompat.setOnActionExpandListener(item,
                new MenuItemCompat.OnActionExpandListener() {
                    @Override
                    public boolean onMenuItemActionCollapse(MenuItem item) {
                        // Do something when collapsed
                        adapter.setFilter(getListClientes());
                        return true; // Return true to collapse action view
                    }

                    @Override
                    public boolean onMenuItemActionExpand(MenuItem item) {
                        // Do something when expanded
                        return true; // Return true to expand action view
                    }
                });
    }

    @Override
    public boolean onQueryTextChange(String newText) {
        final List<sale> filteredModelList = filter(getListClientes(), newText);
        adapter.setFilter(filteredModelList);
        return true;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        return false;
    }


    private List<sale> filter(List<sale> models, String query) {
        query = query.toLowerCase();
        final List<sale> filteredModelList = new ArrayList<>();

        for (sale model : models) {

            if(model.getFantasy() == null){

            }else{

            final String text = model.getFantasy().toLowerCase();
            Log.d("FiltroPreventa", text);
            if (text.contains(query)) {
                filteredModelList.add(model);
            }
            }
        }
        return filteredModelList;
    }

}

File: app/src/main/java/com/friendlypos/distribucion/fragment/DistSelecClienteFragment.kt
package com.friendlysystemgroup.friendlypos.distribucion.fragment

import android.app.Activity
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.widget.SearchView
import androidx.core.view.MenuItemCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.friendlysystemgroup.friendlypos.R
import com.friendlysystemgroup.friendlypos.databinding.FragmentDistSelecClienteBinding
import com.friendlysystemgroup.friendlypos.distribucion.activity.DistribucionActivity
import com.friendlysystemgroup.friendlypos.distribucion.adapters.DistrClientesAdapter
import com.friendlysystemgroup.friendlypos.distribucion.adapters.DistrResumenAdapter
import com.friendlysystemgroup.friendlypos.distribucion.modelo.sale
import com.friendlysystemgroup.friendlypos.modelos.Cliente
import io.realm.Realm
import io.realm.RealmQuery
import io.realm.RealmResults
import io.realm.Sort
import io.realm.internal.SyncObjectServerFacade
import java.util.Locale

class DistSelecClienteFragment : BaseFragment(), SearchView.OnQueryTextListener {
    private var binding: FragmentDistSelecClienteBinding? = null
    private var realm: Realm? = null
    private var adapter: DistrClientesAdapter? = null
    private var adapter2: DistrResumenAdapter? = null
    var activity: DistribucionActivity? = null
    private var clientes: List<Cliente> = listOf()
    private val clientesFilter: MutableList<Cliente> = mutableListOf()

    @Suppress("DEPRECATION")
    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        this.activity = activity as DistribucionActivity
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentDistSelecClienteBinding.inflate(inflater, container, false)
        setHasOptionsMenu(true)
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupViews()
    }

    private fun setupViews() {
        binding?.apply {
            recyclerViewDistrCliente?.layoutManager = LinearLayoutManager(context)
            
            clientes = listaClientes
            clientesFilter.clear()
            clientesFilter.addAll(clientes)
            
            adapter = DistrClientesAdapter(
                context, activity,
                clientesFilter
            )
            adapter2 = DistrResumenAdapter()
            recyclerViewDistrCliente?.adapter = adapter
            
            searchView?.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String): Boolean {
                    return false
                }

                override fun onQueryTextChange(newText: String): Boolean {
                    filterList(newText)
                    return true
                }
            })
        }
    }

    private fun filterList(filterText: String) {
        clientesFilter.clear()
        
        if (filterText.isEmpty()) {
            clientesFilter.addAll(clientes)
        } else {
            val lowerCaseFilter = filterText.lowercase(Locale.getDefault())
            
            clientes.forEach { cliente ->
                val nombre = cliente.nombre?.lowercase(Locale.getDefault()) ?: ""
                val codigo = cliente.codigo?.lowercase(Locale.getDefault()) ?: ""
                
                if (nombre.contains(lowerCaseFilter) || codigo.contains(lowerCaseFilter)) {
                    clientesFilter.add(cliente)
                }
            }
        }
        
        adapter?.setFilter(clientesFilter)
        adapter?.notifyDataSetChanged()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onDetach() {
        super.onDetach()
        activity = null
    }

    val listaClientes: List<Cliente>
        get() {
            val facturaId = activity?.invoiceId ?: return emptyList()
            Log.d("facturaid", facturaId.toString())
            
            val realm = Realm.getDefaultInstance()
            try {
                val clientes: RealmResults<Cliente> = realm.where(Cliente::class.java)
                    .equalTo("isAplicado", false)
                    .equalTo("isDevuelto", false)
                    .findAll()
                
                return realm.copyFromRealm(clientes)
            } finally {
                realm.close()
            }
        }

    override fun updateData() {
        activity?.selecClienteTab = 0
        adapter?.notifyDataSetChanged()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_main, menu)

        val item = menu.findItem(R.id.action_search)
        val searchView = MenuItemCompat.getActionView(item) as SearchView
        searchView.setOnQueryTextListener(this)

        MenuItemCompat.setOnActionExpandListener(
            item,
            object : MenuItemCompat.OnActionExpandListener {
                override fun onMenuItemActionCollapse(item: MenuItem): Boolean {
                    adapter!!.setFilter(this.listaClientes)
                    return true
                }

                override fun onMenuItemActionExpand(item: MenuItem): Boolean {
                    return true
                }
            })
    }

    override fun onQueryTextChange(newText: String): Boolean {
        filterList(newText)
        return true
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        return false
    }

    private fun filter(models: List<sale>, query: String): List<sale> {
        var query = query
        query = query.lowercase(Locale.getDefault())
        val filteredModelList: MutableList<sale> = ArrayList()

        for (model in models) {
            if (model.fantasy == null) {
            } else {
                val text = model.fantasy.lowercase(Locale.getDefault())
                Log.d("FiltroPreventa", text)
                if (text.contains(query)) {
                    filteredModelList.add(model)
                }
            }
        }
        return filteredModelList
    }

    companion object {
        val instance: DistSelecClienteFragment
            get() = DistSelecClienteFragment()
    }
}
